pkg/k8s/custom_controller.go

```diff
@@ -0,0 +1,258 @@
+	dataStore cache.Indexer
+}
+
+// NewCustomController returns a new podController object
+func NewCustomController(clientSet *kubernetes.Clientset, pageLimit int64, namesspace string, converter Converter, resyncPeriod time.Duration,
+	retryOnError bool, eventNotificationChan chan<- GenericEvent, log logr.Logger) *CustomController {
+	c := &CustomController{
+		clientSet:             clientSet,
+		pageLimit:             pageLimit,
+		namespace:             namesspace,
+		converter:             converter,
+		resyncPeriod:          resyncPeriod,
+		retryOnError:          retryOnError,
+		eventNotificationChan: eventNotificationChan,
+		log:                   log,
+	}
+	c.dataStore = cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{NamespaceIndexKey: NamespaceKeyIndexFunc()})
+	c.queue = cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, c.dataStore)
+	return c
+}
+
+// StartController starts the custom controller by doing a list and watch on the specified k8s
+// resource. The controller would store the converted k8s object in the provided indexer. The
+// stop channel should be notified to stop the controller
+func (c *CustomController) StartController(stopChanel <-chan struct{}) {
+	config := &cache.Config{
+		Queue: c.queue,
+		ListerWatcher: newListWatcher(c.clientSet.CoreV1().RESTClient(),
+			c.converter.Resource(), c.namespace, c.pageLimit, c.converter),
+		ObjectType:       c.converter.ResourceType(),
+		FullResyncPeriod: c.resyncPeriod,
+		RetryOnError:     c.retryOnError,
+		Process: func(obj interface{}) error {
+			// from oldest to newest
+			for _, d := range obj.(cache.Deltas) {
+				// Strip down the pod object and keep only the required details
+				convertedObj, err := c.converter.ConvertObject(d.Object)
+				if err != nil {
+					return err
+				}
+				switch d.Type {
+				case cache.Sync, cache.Added, cache.Updated:
+					if old, exists, err := c.dataStore.Get(convertedObj); err == nil && exists {
+						if err := c.dataStore.Update(convertedObj); err != nil {
+							return err
+						}
+						if err := c.notifyChannelOnUpdate(old, convertedObj); err != nil {
+							return err
+						}
+					} else if err == nil && !exists {
+						if err := c.dataStore.Add(convertedObj); err != nil {
+							return err
+						}
+						if err := c.notifyChannelOnCreate(convertedObj); err != nil {
+							return err
+						}
+					} else {
+						return err
+					}
+				case cache.Deleted:
+					if err := c.dataStore.Delete(convertedObj); err != nil {
+						return err
+					}
+					if err := c.notifyChannelOnDelete(convertedObj); err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		},
+	}
+	c.controller = cache.New(config)
+
+	// Run the controller
+	c.controller.Run(stopChanel)
+}
+
+// GetDataStore returns the data store when it has successfully synced with API Server
+func (c *CustomController) GetDataStore() cache.Indexer {
+	// Custom data store, it should not be accessed directly as the cache could be out of sync
```

Comment:
nit, i'm ok with this implementation as it's already used by vpc resource controller. 
However, we should refactor this in the future, as this approach will make controllers to run without have pod synced. (it actually blocks when pod information is accessed).

Reply:
+1 I will take an action item to see if we can refactor to sync the cache first and then start the controllers.